gen/zap.go

```diff
@@ -86,6 +86,27 @@ func (z *zapGenerator) zapEncoder(g Generator, spec compile.TypeSpec) string {
 	panic(root)
 }
 
+// zapTypedefHasGeneratedMarshaler defines if a typedef will have generated code
+// zapcore.ObjectMarshaler or zapcore.ArrayMarshaler.
+func (z *zapGenerator) zapTypedefHasGeneratedMarshaler(g Generator, spec compile.TypeSpec) bool {
+	return z.zapEncoder(g, spec) == "Array" || z.zapEncoder(g, spec) == "Object"
```

Comment:
nit: can use switch here. Will also reduce number of zapEncoder calls.

```suggestion
	switch z.zapEncoder(g, spec) {
	case "Array", "Object":
		return true
	default:
		return false
	}
```